python/benchmark/gen_data_distributed.py

```diff
@@ -502,6 +513,286 @@ def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+
+        rows = self.num_rows
+        cols = self.num_cols
+        assert self.args is not None
+        num_partitions = self.args.output_num_files
+
+        # Set num_partitions to Spark's default if output_num_files is not provided.
+        if num_partitions is None:
+            num_partitions = spark.sparkContext.defaultParallelism
+
+        # Retrieve input params or set to defaults.
+        seed = params["random_state"]
+        generator = np.random.RandomState(seed)
+        bias = params.get("bias", 0.0)
+        noise = params.get("noise", 0.0)
+        shuffle = params.get("shuffle", True)
+        n_informative = params.get("n_informative", 10)
+        n_targets = params.get("n_targets", 1)
+        use_gpu = params.get("use_gpu", False)
+        logistic_regression = params.get("logistic_regression", False)
+        density = params.get("density", 0.1)
+        redundant_cols = params.get("redundant_cols", 0)
+        orig_cols = cols - redundant_cols
+
+        # Check for valid redundant columns
+        if redundant_cols > 0 and redundant_cols / (cols + redundant_cols) > density:
+            logging.warning(
+                "Redundant columns would break density property, setting to zero instead"
+            )
+            redundant_cols = 0
+            orig_cols = cols
+
+        # Generate ground truth upfront.
+        ground_truth = np.zeros((cols, n_targets))
+        ground_truth[:n_informative, :] = 100 * generator.uniform(
+            size=(n_informative, n_targets)
+        )
+
+        if shuffle:
+            # Shuffle feature indices upfront.
+            col_indices = np.arange(cols)
+            generator.shuffle(col_indices)
+            ground_truth = ground_truth[col_indices]
+
+        # Create different partition seeds for sample generation.
+        random.seed(params["random_state"])
+        seed_maxval = 100 * num_partitions
+        partition_seeds = random.sample(range(1, seed_maxval), num_partitions)
+
+        # UDF for distributed generation of X and y.
+        def make_sparse_regression_udf(
+            iter: Iterable[pd.DataFrame],
+        ) -> Iterable[pd.DataFrame]:
+            use_cupy = use_gpu
+            if use_cupy:
+                try:
+                    import cupy as cp
+                except ImportError:
+                    use_cupy = False
+                    logging.warning("cupy import failed; falling back to numpy.")
+
+            partition_index = pyspark.TaskContext().partitionId()
+
+            # Get #rows in charge
+            num_rows_per_partition = 0
+            start = -1
+
+            dfs = []
+            for pdf in iter:
+                num_rows_per_partition += pdf.shape[0]
+                dfs.append(pdf)
+                if start == -1:
+                    start = pdf["id"][0]
+
+            # Add in redundant cols of linear combinations of generated random sparse cols
+            if redundant_cols > 0:
```

Comment:
Does this PR support redundant_cols != 0? If so, will need to test it in the test functions. If no, will need to move the parameter to the next PR.